Added NLP: Basic text classification with JAX & Flax (#95)
jax-ml/jax-ai-stack: docs/JAX_basic_text_classification.md

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.15.2
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

# Basic Text classification with JAX & FLAX

> In this tutorial we learn how to perform text classification from raw text data and train a basic 1D Convnet to perform sentiment analysis using JAX. This tutorial is originally inspired by ["Text classification from scratch with Keras"](https://keras.io/examples/nlp/text_classification_from_scratch/#build-a-model) tutorial.

Reviewer: ```suggestion
In this tutorial we learn how to perform text classification from raw text data and train a basic 1D Convnet to perform sentiment analysis using JAX. This tutorial is originally inspired by ["Text classification from scratch with Keras"](https://keras.io/examples/nlp/text_classification_from_scratch/#build-a-model).
```

> We are using IMDB movie review dataset to classify the review to "positive" and "negative" classes. We implement from scratch a simple model using Flax, train it and compute metrics on test set.

Reviewer: ```suggestion
We will use the IMDB movie review dataset to classify the review to "positive" and "negative" classes. We implement from scratch a simple model using Flax, train it and compute metrics on the test set.
```

+++

## Setup

We will be using the following packages in this tutorial:
- [Tiktoken](https://github.com/openai/tiktoken) to tokenize the raw text
- [Grain](https://github.com/google/grain) for efficient data loading and batching

> - [tqdm](https://tqdm.github.io/) for a progress bar in the training

Reviewer: nit:

```suggestion
- [tqdm](https://tqdm.github.io/) for a progress bar in the training step
```

Reviewer: Just ignore then, thanks :)

```{code-cell} ipython3
!pip install grain tiktoken tqdm
```

### Load the data: IMDB movie review sentiment classification

Let us download the dataset and briefly inspect the structure. We will be using only two classes: "positive" and "negative" for the sentiment analysis.

```{code-cell} ipython3
!rm -rf /tmp/data/imdb
!mkdir -p /tmp/data/imdb
!wget https://ai.stanford.edu/~amaas/data/sentiment/aclImdb_v1.tar.gz -O /tmp/data/imdb/aclImdb_v1.tar.gz
!cd /tmp/data/imdb/ && tar -xf aclImdb_v1.tar.gz
```

```{code-cell} ipython3
!echo "Number of positive samples in train set:"
!ls /tmp/data/imdb/aclImdb/train/pos | wc -l
!echo "Number of negative samples in train set:"
!ls /tmp/data/imdb/aclImdb/train/neg | wc -l
!echo "Number of positive samples in test set:"
!ls /tmp/data/imdb/aclImdb/test/pos | wc -l
!echo "Number of negative samples in test set:"
!ls /tmp/data/imdb/aclImdb/test/neg | wc -l
!echo "First 10 files with positive samples in train/test sets:"
!ls /tmp/data/imdb/aclImdb/train/pos | head
!ls /tmp/data/imdb/aclImdb/test/pos | head
!echo "Display a single positive sample:"
!cat /tmp/data/imdb/aclImdb/train/pos/6248_7.txt
```

> Next, we create the dataset Python class to read samples from the disk, use [Tiktoken](https://github.com/openai/tiktoken) to encode raw text into tokens and use [Grain](https://github.com/google/grain) for efficient data loading and batching.

Reviewer: ```suggestion
Next, we will:
- create the dataset Python class to read samples from the disk;
- use [Tiktoken](https://github.com/openai/tiktoken) to encode raw text into tokens; and
- use [Grain](https://github.com/google/grain) for efficient data loading and batching.
```

        assert self.path.exists()

        pos_texts = list((self.path / "pos").glob("*.txt"))
        neg_texts = list((self.path / "neg").glob("*.txt"))
        self.text_files = pos_texts + neg_texts
        assert len(self.text_files) > 0
        # Label 0 for Positive comments
        # Label 1 for Negative comments
        self.labels = [0] * len(pos_texts) + [1] * len(neg_texts)

    def __len__(self) -> int:
        return len(self.text_files)

    def read_text_file(self, path: str | Path) -> str:
        with open(path, "r") as handler:
            lines = handler.readlines()
        return "\n".join(lines)

    def __getitem__(self, index: int) -> tuple[str, int]:
        label = self.labels[index]
        text = self.read_text_file(self.text_files[index])
        return {"text": text, "label": label}

root_path = Path("/tmp/data/imdb/aclImdb/")
train_dataset = SentimentAnalysisDataset(root_path / "train")
test_dataset = SentimentAnalysisDataset(root_path / "test")

print("- Number of samples in train and test sets:", len(train_dataset), len(test_dataset))
print("- First train sample:", train_dataset[0])
print("- First test sample:", test_dataset[0])
```

> Now, we can create string to tokens preprocessing transformation and set up data loaders. We are going to use the GPT-2 tokenizer via [Tiktoken](https://github.com/openai/tiktoken).

Reviewer: This reads better to me, but it's only a suggestion:
```suggestion
Now, we can create a string-to-tokens preprocessing transformation and set up data loaders. We are going to use the GPT-2 tokenizer via [Tiktoken](https://github.com/openai/tiktoken).
```

```{code-cell} ipython3
import numpy as np

import tiktoken
import grain.python as grain

tokenizer = tiktoken.get_encoding("gpt2")

class TextPreprocessing(grain.MapTransform):
    def __init__(self, tokenizer, max_length: int = 256):
        self.tokenizer = tokenizer
        self.max_length = max_length

    def map(self, data):
        text = data["text"]
        encoded = self.tokenizer.encode(text)
        # Cut to max length
        encoded = encoded[:self.max_length]
        # Pad with zeros if needed
        encoded = np.array(encoded + [0] * (self.max_length - len(encoded)))
        return {
            "text": encoded,
            "label": data["label"],
        }
```

> 

Reviewer: If you are using two separate code cells for these actions, I would keep the grain import in the next cell so they are "self-contained" (they are not really self-contained, but it makes for a better cognitive unit)

    worker_buffer_size=2,                  # Count of output batches to produce in advance per worker
    operations=[
        TextPreprocessing(tokenizer, max_length=max_length),
        grain.Batch(train_batch_size, drop_remainder=True),
    ]
)

test_loader = grain.DataLoader(
    data_source=test_dataset,
    sampler=test_sampler,                  # Sampler to determine how to access the data
    worker_count=4,                        # Number of child processes launched to parallelize the transformations among
    worker_buffer_size=2,                  # Count of output batches to produce in advance per worker
    operations=[
        TextPreprocessing(tokenizer, max_length=max_length),
        grain.Batch(test_batch_size),
    ]
)
```

```{code-cell} ipython3
train_batch = next(iter(train_loader))
```

```{code-cell} ipython3
print("Train encoded text batch info:", type(train_batch["text"]), train_batch["text"].shape, train_batch["text"].dtype)
print("Train labels batch info:", type(train_batch["label"]), train_batch["label"].shape, train_batch["label"].dtype)
```

```{code-cell} ipython3

> print("Batch data:", train_batch["text"][:2, :12], train_batch["label"][:2])

Reviewer: Can you leave a tiny note about why we are printing this, and what we should expect to see?

```

## Model for text classification

> We choose a simple 1D convnet to classify the text. The first layer of the model transforms input tokens into float features using an embedding layer (`nnx.Embed`), then they are encoded further with convolutions. Finally, we classify encoded features using a fully-connected layers.

Reviewer: ```suggestion
We choose a simple 1D convnet to classify the text. The first layer of the model transforms input tokens into float features using an embedding layer (`nnx.Embed`), then they are encoded further with convolutions. Finally, we classify encoded features using fully-connected layers.
```

            out_features=hidden_dim,
            kernel_size=conv_ksize,
            strides=conv_ksize // 2,
            rngs=rngs,
        )
        self.lnorm2 = nnx.LayerNorm(hidden_dim, rngs=rngs)

        self.fc1 = nnx.Linear(hidden_dim, hidden_dim, rngs=rngs)
        self.fc2 = nnx.Linear(hidden_dim, num_classes, rngs=rngs)

    def __call__(self, x: jax.Array) -> jax.Array:
        # x.shape: (N, max_length)
        x = self.token_embedding(x)
        x = self.dropout(x)   # x.shape: (N, max_length, embed_dim)

        x = self.conv1(x)
        x = self.lnorm1(x)
        x = self.activation_layer(x)
        x = self.conv2(x)
        x = self.lnorm2(x)
        x = self.activation_layer(x)  # x.shape: (N, K, hidden_dim)

        x = nnx.max_pool(x, window_shape=(x.shape[1], ))  # x.shape: (N, 1, hidden_dim)
        x = x.reshape((-1, x.shape[-1]))  # x.shape: (N, hidden_dim)

        x = self.fc1(x)  # x.shape: (N, hidden_dim)
        x = self.activation_layer(x)
        x = self.dropout(x)
        x = self.fc2(x)  # x.shape: (N, 2)

        return x

x = jnp.ones((4, max_length), dtype="int32")
module = TextConvNet(vocab_size)
y = module(x)

> print(y.shape)

Reviewer: Why is this shape informative here?

Author: To ensure that we output expected shape: (N, num_classes), updated to
```
print("Prediction shape (N, num_classes): ", y.shape)
```

def train_one_epoch(epoch):
    model.train()  # Set model to the training mode: e.g. update batch statistics
    with tqdm.tqdm(
        desc=f"[train] epoch: {epoch}/{num_epochs}, ",
        total=train_total_steps,
        bar_format=bar_format,
        leave=True,
    ) as pbar:
        for batch in train_loader:
            loss = train_step(model, optimizer, batch)
            train_metrics_history["train_loss"].append(loss.item())
            pbar.set_postfix({"loss": loss.item()})
            pbar.update(1)

def evaluate_model(epoch):
    # Compute the metrics on the train and val sets after each training epoch.
    model.eval()  # Set model to evaluation model: e.g. use stored batch statistics

    eval_metrics.reset()  # Reset the eval metrics
    for test_batch in test_loader:
        eval_step(model, test_batch, eval_metrics)

    for metric, value in eval_metrics.compute().items():
        eval_metrics_history[f'test_{metric}'].append(value)

    print(f"[test] epoch: {epoch + 1}/{num_epochs}")
    print(f"- total loss: {eval_metrics_history['test_loss'][-1]:0.4f}")
    print(f"- Accuracy: {eval_metrics_history['test_accuracy'][-1]:0.4f}")
```

> Let us now start the training

Reviewer: Just a suggestion:
```suggestion
Now, we can start the training. Note that this step can take quite a while to run, depending on the number of epochs you are using. 
```

Author: Well, this is not exactly true, in this case, whole training takes 2 minutes. The model and the dataset are rather small.

Reviewer: Interesting! For me it took something like 40 mins in colab.

```{code-cell} ipython3
for epoch in range(num_epochs):
    train_one_epoch(epoch)
    evaluate_model(epoch)
```

> Let us visualize collected metrics

Reviewer: ```suggestion
Let's visualize the collected metrics:
```

```{code-cell} ipython3
import matplotlib.pyplot as plt

plt.plot(train_metrics_history["train_loss"], label="Loss value during the training")
plt.legend()
```

```{code-cell} ipython3
fig, axs = plt.subplots(1, 2, figsize=(10, 10))
axs[0].set_title("Loss value on test set")
axs[0].plot(eval_metrics_history["test_loss"])
axs[1].set_title("Accuracy on test set")
axs[1].plot(eval_metrics_history["test_accuracy"])
```

We can observe that the model starts overfitting after the 5-th epoch and the best accuracy it could achieve is around 0.87. Let us also check model's predictions on the test data:

```{code-cell} ipython3
data = test_dataset[10]
```

```{code-cell} ipython3
text_processing = TextPreprocessing(tokenizer, max_length=max_length)
processed_data = text_processing.map(data)
```

```{code-cell} ipython3
model.eval()
preds = model(processed_data["text"][None, :])
preds, preds.argmax(axis=-1)
```

```{code-cell} ipython3
print("Text:\n", data["text"])
print("Expected label:", data["label"])
print("Predicted label:", preds.argmax(axis=-1).item())
```

> 

Reviewer: Is it possible to include a quick summary/interpretation of what we are seeing in these labels?

---
jupytext:
  formats: ipynb,md:myst
  text_representation:
    extension: .md
    format_name: myst
    format_version: 0.13
    jupytext_version: 1.15.2
kernelspec:
  display_name: Python 3 (ipykernel)
  language: python
  name: python3
---

> # Basic Text classification with JAX & FLAX

Reviewer 2: Let's say "1D Convnet for basic text classification"

```

```{code-cell} ipython3
text_processing = TextPreprocessing(tokenizer, max_length=max_length)
processed_data = text_processing.map(data)
model.eval()
preds = model(processed_data["text"][None, :])
pred_label = preds.argmax(axis=-1).item()
confidence = nnx.softmax(preds, axis=-1)

print("- Text:\n", data["text"])
print("")
print(f"- Expected review sentiment: {'positive' if data['label'] == 0 else 'negative'}")
print(f"- Predicted review sentiment: {'positive' if pred_label == 0 else 'negative'}, confidence: {confidence[0, pred_label]:.3f}")
```

```{code-cell} ipython3
data = test_dataset[20]
```

```{code-cell} ipython3
text_processing = TextPreprocessing(tokenizer, max_length=max_length)
processed_data = text_processing.map(data)
model.eval()
preds = model(processed_data["text"][None, :])
pred_label = preds.argmax(axis=-1).item()
confidence = nnx.softmax(preds, axis=-1)

print("- Text:\n", data["text"])
print("")
print(f"- Expected review sentiment: {'positive' if data['label'] == 0 else 'negative'}")
print(f"- Predicted review sentiment: {'positive' if pred_label == 0 else 'negative'}, confidence: {confidence[0, pred_label]:.3f}")
```

## Further reading

> - Model checkpointing and exporting using [Orbax](https://orbax.readthedocs.io/en/latest/)

Reviewer 2: It would be nice to have some sort of conclusion text here – summarize what we did and what the next steps may be for someone interested in going deeper.

Author: Yes, I wanted to add links on other NLP tutorials doing text classification with transformers once we have them.
I'll add a short summary on what we did.

> ```{code-cell} ipython3

Reviewer 2: Remove empty cell
